### ISSUE TITLE
Add participants per room metrics

### ISSUE THREAD

**Issue by vol-pi** [greenstatic/bigbluebutton-exporter#9]
First of all: thank you for your great work on this exporter. Your release enormously helped us to roll out a bigger Scalelite > BBB installation for a university.

However, to better understand the load pattern we need a more detailed look at the room participant distribution on each server. I've created a pull request with some additional changes here #8 

Best regards

**Comment by greenstatic** [comment]
I'm delighted that you managed to scale up BBB at your university. We are currently looking into Scalelite to scale our BBB installation down the road.

I'll continue [my comment](https://github.com/greenstatic/bigbluebutton-exporter/pull/8#issuecomment-612161133) from your PR request here.

Although the feature request is completely valid and would greatly benefit the admin responsible for the BBB installation, I am worried about the unbounded nature of the meetings label.

Excerpt from [Prometheus documentation](https://prometheus.io/docs/practices/naming/#labels):
>  CAUTION: Remember that every unique combination of key-value label pairs represents a new time series, which can dramatically increase the amount of data stored. Do not use labels to store dimensions with high cardinality (many different label values), such as user IDs, email addresses, or other unbounded sets of values.

Meaning that in general one should have a "low" number of bounded label values - some say a general rule of thumb is 10 per scraping target.

A small back of the envelope calculation for an educational institution:
30 daily meetings x 21 avg. working days in a month x 4 months per semester x 2 semesters = **5040 unique meetings per year**.

Now the 30 daily meetings is a relatively small figure for a bigger edu. institution and this does not count the situation where students are allowed to create rooms themselves. So this figure really depends on a per-case situation.

But it gives a valuable insight that the label is "high cardinal".


With @matjazp we came up with a couple of ideas:
1. We keep it as you implemented it - room id as the label value

Pros:
- we have precise metrics
- globally unique, so no issues when scaling with scalelite

Cons:
- the label value is unbounded resulting in possible issues down the road with Prometheus
- loads of time series

---

2. Since in general we are interested in distinct rooms at a moment in time and not in particular with a unique room in a long (1 week) time range we could have a bounded value for a room - e.g. room 1 - 500. When scraping the API data we deterministically based on the the meetings metadata choose a room from the range (1-500). This would provide an upper bound on the number of distinct label values.

Pros:
- label value upper bound
- we can track distinct rooms using Prometheus

Cons:
- confusing long term queries since a meeting from a week/month ago could/will fall into the same "bucket" (same label) as another recent meeting metric
- hackish and not a common design decision with metrics
- race condition when two concurrent meetings collide into the same "bucket" (range e.g. 1-500)

---

3. We include such functionality in the logging system (i.e. ELK stack)

Pros:
- we do not have the label cardinality issue

Cons:
- split metric like data into two systems (Prometheus and e.g. ELK)
- we require a logging system setup :)

---
I am personally more for option 3. But I am pretty sure there are alternative ideas how to mitigate this issue that we haven't thought of yet. On the other hand, considering kubernetes' nginx-ingress exposes metrics which contain the controller's container name, it is possible that the "high cardinality" is a nonissue for our use case. (Although to be fair, how often does nginx-ingress-controller in k8s change containers? Probably not as many times in a day than the typical number of meetings in BBB).


What do you think @vol-pi , is the unbounded nature of the label values a nonissue for you or do you have some other idea how to solve this issue?

**Comment by vol-pi** [comment]
Thank you for your elaborated answer. Regarding your concerns towards capacity planing:

As of now, I have 3 exporters running: Your `bbb_exporter`, the `node_exporter` and `process_exporter` (for freeswitch and kurento). The `node_exporter` alone has around 16000 metrics (`count({__name__=~"node_.+"})`)

The Prometheus documentation has a passage about [capacity planning](https://prometheus.io/docs/prometheus/latest/storage/#operational-aspects). They suggest the following formula: 

```
needed_disk_space = retention_time_seconds * ingested_samples_per_second * bytes_per_sample
```

Keeping all metrics in Prometheus over the course of a year would result in a huge amount of storage. Therefore some kind of long term storage and downsampling becomes a necessity. In my opinion it is good idea to support as much metrics as possible and leave it to the user to enable them and consider their long term storage.

My suggestions would be to choose option 1) and make the room metrics optional, configurable via environment variable. I'd update my pull request if you like.

**Comment by greenstatic** [comment]
Great protip regarding process_exporter for monitoring freeswitch and kurento. I wish I knew that before, when our kurento processes skyrocketed to 100% once per week at random (before we updated to 2.2.5). It would make alerting much easier.

I think you calculated your metrics wrong, you probably have multiple instances that are being scraped and not all node_exporter metrics are prefixed with `node_`. 
When I run the following query: `count({job="<node_exporter job name>"}) by (instance)`
I get I only get 1289 metrics for each bbb server with node_exporter. 

I agree with you that it is a good idea to support as many metrics as possible and I really like your idea with the room metric being optional and enabling it via ENV variable.

Before we go with option 1, I would just like to throw the idea of using maybe a [histogram](https://prometheus.io/docs/concepts/metric_types/#histogram) instead of a gauge. The API response times are collected in this way. Maybe we could use a similar technique for participants per room and have a single histogram? This would enable us to get rid of the room id label. Haven't though it over completely yet but what do you think?

**Comment by greenstatic** [comment]
I have implemented a variation of the suggested metric called `bbb_room_participants_bucket`. This is a histogram with default bucket sizes of ` 0,1,5,15,30,60,90,120,150,200,250,300,400,500` where we count the number of rooms and place them into the appropriate bucket. See [Prometheus documentation](https://prometheus.io/docs/concepts/metric_types/#histogram) regarding details on the histogram metric type.

You can also override the bucket sizes using the ENV variable `ROOM_PARTICIPANTS_CUSTOM_BUCKETS` - see documentation for details.

Bellow is a screenshot of the custom Grafana panel showing the histogram metric:

![Screenshot from 2020-05-07 20-52-23](https://user-images.githubusercontent.com/14580434/81333723-49c6f280-90a5-11ea-83eb-6e821e947e81.png)

Unfortunately I don't have much data as of now so I will have to wait until tomorrow to gather more data and see if everything is in order before merging into master.

This metric is being developed in the [participant-histogram](https://github.com/greenstatic/bigbluebutton-exporter/tree/participant-histogram) branch (this includes the documentation that goes along documenting the new metric).

You can try out a preview build of this branch by pulling the docker image: `greenstatic/bigbluebutton-exporter:v0.3.0-participant-histogram`. This docker tag will not pin you to a specific version, but will follow the latest build of the branch `participant-histogram` and will be removed once the branch is merged into master.

I would be interested in your input @vol-pi  and @genofire if this solves your issue. I prefer this implementation compared to PR #8 since we do not burden Prometheus with the high cardinality room id labels. Such information would be more appropriate to store in the logging system than metrics system. I'm planning on integrating BBB into ELK stack  where such functionality could be implemented. I do not however have an ETA when I'll manage to get around to implement that.

The Grafana panel settings for the screenshot above is:
```json
{
  "datasource": null,
  "cacheTimeout": null,
  "cards": {
    "cardPadding": null,
    "cardRound": null
  },
  "color": {
    "cardColor": "#b4ff00",
    "colorScale": "linear",
    "colorScheme": "interpolateGnBu",
    "exponent": 0.5,
    "mode": "spectrum"
  },
  "dataFormat": "tsbuckets",
  "description": "Buckets (Y axis) represents the number of participants while the count in each bucket represents the number of rooms with the bucket's size value participants. ",
  "gridPos": {
    "h": 10,
    "w": 14,
    "x": 0,
    "y": 27
  },
  "heatmap": {},
  "hideZeroBuckets": true,
  "highlightCards": true,
  "id": 44,
  "legend": {
    "show": true
  },
  "links": [],
  "pluginVersion": "6.7.2",
  "reverseYBuckets": false,
  "targets": [
    {
      "expr": "bbb_room_participants_bucket",
      "format": "heatmap",
      "instant": false,
      "interval": "",
      "legendFormat": "{{le}}",
      "refId": "A"
    }
  ],
  "timeFrom": null,
  "timeShift": null,
  "title": "Heatmap of Room Participants",
  "tooltip": {
    "show": true,
    "showHistogram": false
  },
  "type": "heatmap",
  "xAxis": {
    "show": true
  },
  "xBucketNumber": null,
  "xBucketSize": null,
  "yAxis": {
    "decimals": null,
    "format": "short",
    "logBase": 1,
    "max": null,
    "min": null,
    "show": true,
    "splitFactor": null
  },
  "yBucketBound": "auto",
  "yBucketNumber": null,
  "yBucketSize": null
}
```

**Comment by genofire** [comment]
It is really nice, what do you thought about video, voice and listen ?

**Comment by greenstatic** [comment]
If I understand correctly, you would like histograms for video, voice and listeners of each room. Yeah I could do that.

**Comment by genofire** [comment]
Oh yes, thats what i mean.
Otherwise we, are running it and it is in current state already really nice :)

**Comment by greenstatic** [comment]
Thats great to hear! I’ll check in the evening on our end if everything is good and then implement the additional histograms tonight. I’ll probably merge this during the weekend into master.

**Comment by greenstatic** [comment]
[Latest commit](https://github.com/greenstatic/bigbluebutton-exporter/commit/e5055641d6e02fc871419107133f5bc810cc3194) in branch [participant-histogram](https://github.com/greenstatic/bigbluebutton-exporter/compare/participant-histogram) has besides the histogram:
```
bbb_room_participants_bucket
```

also:
```
bbb_room_listeners_bucket
bbb_room_voice_participants_bucket
bbb_room_video_participants_bucket
```

It is available (as a preview) in the latest release of the docker image: `greenstatic/bigbluebutton-exporter:v0.3.0-participant-histogram`
